### PR TITLE
Fixed epsilon decay in dqn example

### PR DESCRIPTION
**Pull request by carsondenison**
# Before submitting

- I raised this issue on github, issue https://github.com/PyTorchLightning/pytorch-lightning/discussions/10883. The user @awaelchli said to fork the repo, clone it to my local machine, make my changes, push to my repo, and then submit a pull request.
- This is my first attempt to contribute to an open-source project, so please let me know if I've done anything incorrectly.
- I could not find where the unit tests are kept for the example colab notebooks. I would be happy to upload my unit tests for the get_epsilon method, but I don't know where to put them in the codebase.

## What does this PR do?

This fixes issue  https://github.com/PyTorchLightning/pytorch-lightning/discussions/10883, where there was an incorrect version of epsilon decay for the epsilon-random policy of a DQN. The original code has a single `train_step` with `self.hparams.eps_start` and then immediately switches to `self.hparams.eps_end`. The intended behavior is to linearly decrease epsilon from `self.haparms.eps_start` to `self.hparams.eps_end` over the first `self.hparams.eps_last_frame` steps. I wrote a small function `get_epsilon` which fixes this logic and returns the correct epsilon.

I have also made a few minor changes on other lines, because the code would not run on my local machine without these changes. Specifically, the type hint for the `__iter__` method of the `RLDataset` class was a Tuple, and should be an Iterator[Tuple], because it returns a generator of tuples representing (state, action, reward, done, new_state). Additionally, on line 264 (formerly 276), I got an error that the index for the `gather()` function must be of type `int64`, so I cast the `actions` tensor to type long. Finally, I added logging of `self.episode_reward` and `epsilon`, so that I could see that a) it still learned successfully, and b) my intended changes to epsilon were working as intended.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Sure did!

### CODE REVIEW

File: lightning_examples/reinforce-learning-DQN/dqn.py [hunk 1]
```diff
@@ -284,6 +284,12 @@ def dqn_mse_loss(self, batch: Tuple[Tensor, Tensor]) -> Tensor:
 
         return nn.MSELoss()(state_action_values, expected_state_action_values)
 
+    def get_epsilon(self, start: int, end: int, frames: int) -> float:
+        if self.global_step > frames:
+            return end
+        else:
+            return start - (self.global_step / frames) * (start - end)
```
Comment (Borda) [inline lightning_examples/reinforce-learning-DQN/dqn.py:291]:
```suggestion
        return start - (self.global_step / frames) * (start - end)
```

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: lightning_examples/reinforce-learning-DQN/dqn.py [hunk 2]
```diff
@@ -363,7 +367,7 @@ def get_device(self, batch) -> str:
 
 trainer = Trainer(
     gpus=AVAIL_GPUS,
-    max_epochs=200,
+    max_epochs=400,
```
Comment (Borda) [inline lightning_examples/reinforce-learning-DQN/dqn.py:370]:
```suggestion
    max_epochs=200,
```
is this needed? there is some timeout issue if this takes too long...

File: lightning_examples/reinforce-learning-DQN/.meta.yml [hunk 3]
```diff
@@ -13,6 +13,10 @@ description: |
   2. Handle unsupervised learning by using an IterableDataset where the dataset itself is constantly updated during training
   3. Each training step carries has the agent taking an action in the environment and storing the experience in the IterableDataset
 requirements:
+  - torch>=1.6, <1.9
```
Comment (Borda) [inline lightning_examples/reinforce-learning-DQN/.meta.yml:16]:
```suggestion
```
this is in defaults

File: lightning_examples/reinforce-learning-DQN/.meta.yml [hunk 4]
```diff
@@ -1,7 +1,7 @@
 title: How to train a Deep Q Network
 author: PL team
 created: 2021-01-31
-updated: 2021-06-17
+updated: 2021-12-03
 license: CC BY-SA
 build: 2
```
Comment (Borda) [inline lightning_examples/reinforce-learning-DQN/.meta.yml:6]:
```suggestion
build: 1
```

File: lightning_examples/reinforce-learning-DQN/dqn.py (hunk 2 above)
Comment (carsondenison) [reply lightning_examples/reinforce-learning-DQN/dqn.py]: Is this why the code initially failed the automated tests? Specifically, it failed code block 8. This is the block that calls trainer.fit(model). I couldn't figure out how to see why it was failing the test, so I wasn't sure what to change.

In general, how do I tell what the error is when I push code that fails one of the automatic tests?
